### PR TITLE
Change how key export works, and add key import functionality as well as key match

### PR DESCRIPTION
**Pull request by simo5**
This are various fixes to small things I had in my tree and to recent issues reported in #265 and #266 

Note that I changed how export work, including an attempt to re-enable direct export from store and then again disable it.

### CODE REVIEW

**Comment by fabled** [comment]
Tested and things don't fully work with the latest commit disabling store export. The things that fails without is that when I get matching public key from store against same openssl public key using the `EVP_PKEY_eq` fails.

**Comment by simo5** [comment]
Are these all RSA keys?
I am aware that EC keys are not working yet.

If not can you describe a simple reproducer (or contribute a test that is a simple reproducer) ?

**Comment by simo5** [comment]
I think EVP_PKEY_eq fails because I did not implement the match() function.

It works when store export is available, because when that happens the public key is exported and imported into the default provider instead, and the default provider implements the match function.

I will add implementation of a match function and EVP_PKEY_eq test to match.

**Comment by simo5** [comment]
I also now know why ECDH fails.
It is because allowing to export from store exports the public key to the default provider, then later openssl tries to import the key back into the pkcs11-provider because it needs it to perform the ECDH using the private key part that is not exportable, but importing keys is not yet supported because I did not think of tying to import keys into a session, and importing to the token is problematic as many tokens do not support that or would cause littering he token with ephemeral keys ...

**Comment by fabled** [comment]
> Are these all RSA keys? I am aware that EC keys are not working yet.

Yes, testing first with RSA keys. Then will move on to EC keys too.

> If not can you describe a simple reproducer (or contribute a test that is a simple reproducer) ?

Sure, I can add a small test case for this. Not sure yet when I get the time slot. Possibly this week, or hopefully by latest early next week. If you get to it first, that's also ok.

**Comment by simo5** [comment]
Nice I will steal it into the PR then

**Comment by simo5** [comment]
So I use this test but it does not reproduce your case.
With this code what happens is that OpenSSL imports the public key into the default provider, so the comparison is happening there ...
I can force keys to not be exportable at all, will see if that gives failures with current code.

**Comment by simo5** [comment]
Forcing keys to stay on the token make them fail without a ec_match function in the provider and succeed when ec_match is available.

**Comment by fabled** [comment]
My commit is in top of the last working commit of this PR. Rebasing it on top of the latest commit in the PR should produce the failure.

**Comment by simo5** [comment]
> My commit is in top of the last working commit of this PR. Rebasing it on top of the latest commit in the PR should produce the failure.

I realized that I had tested it with the patch to remove export disabled.
But that turned out to be a good thing so I could test the matching function.

**Comment by fabled** [comment]
Also, it seems that the public key export works different depending if commit 1ca4a1758d402e614667156d8bbd3c0cc6c458c0 is present or not. With the commit present, the equality check fails because it's unable to load the RSA public key from the `.pem` file.

The difference seems to be that with the commit (store export blocked), the `.pem` file contains `-----BEGIN RSA PUBLIC KEY-----                                                  
` block. And without the commit there's a `-----BEGIN PUBLIC KEY-----` block. Also the whole ASN1 structure is different. I suspect this is also a regression?

**Comment by simo5** [comment]
It is a side effect of exporting to the default provider or not. I am working on various things to allow us to decide what to do.
One area is working on allowing openssl to import keys into the pkcs11-provider... but in the end muy intention is still to block store exporting as it has more side effects than not blocking it.

**Comment by simo5** [comment]
I also filed https://github.com/openssl/openssl/issues/21596

**Comment by simo5** [comment]
@fabled the current set of commits makes all tests pass, but that does not mean everything I did is enough for the way bind uses the API.

Can you please recheck RSA (and maybe even EC) with this patchset applied?

**Comment by fabled** [comment]
> Can you please recheck RSA (and maybe even EC) with this patchset applied?

With checkout at commit e763a88ecdb9fcdbae2d52ffe981272e6ca15294 and `pkcs11-module-load-behavior = early` (still required), the first set of BIND RSA and EC tests PASS. Thank you!

I'll work next with the second set of tests. But I would expect those to pass also, as they should not introduce different paths to openssl, just different paths inside bind.

**Comment by simo5** [comment]
@Jakuje can you help me figuring out why address sanitizer blows up on Debian on a single test with Softhsm ?

**Comment by Jakuje** [comment]
hmm. Sounds like an issue in debian side ... we use Sid, which is rolling release and it might have some bugs. Does the Bookworm/`debian:latest` already have OpenSSL 3.x so we could use more stable release?

I just rerun the test locally in container and it worked ok for me.  Can you try to rerun it in github or try locally?

**Comment by simo5** [comment]
@Jakuje rerunning does not help.
I guess I will ignore this test for now, and either disable asan on debian in a separate PR or find out if there is a more stable debian version with OpenSSL 3 we can use.

**Comment by Jakuje** [comment]
Problem is that I can not reproduce the issue locally in container. Sorry ...

> if there is a more stable debian version with OpenSSL 3 we can use.

Bookworm has 3.0: https://packages.debian.org/bookworm/openssl and should be more stable -- I would give it a try in separate PR. Will need a modification to the debuginfo repo though.

File: src/objects.c [hunk 1]
```diff
@@ -1799,3 +1813,748 @@ int p11prov_obj_export_public_ec_key(P11PROV_OBJ *obj, OSSL_CALLBACK *cb_fn,
...
+    val = ec_params->pValue;
+    group = d2i_ECPKParameters(NULL, &val, ec_params->ulValueLen);
+    if (!group) {
+        ret = RET_OSSL_ERR;
+        goto done;
+    }
+
+    x = BN_CTX_get(bnctx);
+    y = BN_CTX_get(bnctx);
+    if (!y) {
```
Comment (Jakuje) [inline src/objects.c:1922]: missing null check for `x`.

Comment (simo5) [reply src/objects.c:1922]: From openssl man page directly:
`Once BN_CTX_get() has failed, the subsequent calls will return NULL as well, so it is sufficient to check the return value of the last BN_CTX_get() call.`

File: tests/tbasic [hunk 2]
```diff
@@ -152,6 +151,25 @@ fi
 OPENSSL_CONF=${ORIG_OPENSSL_CONF}
 rm -f "${OPENSSL_CONF}.nopin"
```
[the remark this answers is not in the document]
Comment (simo5) [reply tests/tbasic]: Fixed and rebased the relevant commit.

File: src/objects.c [hunk 3]
```diff
@@ -1799,3 +1813,748 @@ int p11prov_obj_export_public_ec_key(P11PROV_OBJ *obj, OSSL_CALLBACK *cb_fn,
...
+        P11PROV_raise(key->ctx, CKR_KEY_INDIGESTIBLE,
+                      "Failure to encode EC point to DER");
+        return CKR_KEY_INDIGESTIBLE;
+    }
+    key->attrs[key->numattrs].type = CKA_EC_POINT;
+    key->attrs[key->numattrs].pValue = der;
+    key->attrs[key->numattrs].ulValueLen = len;
+    key->numattrs++;
+
+    /* TODO: EC_PARAMS */
```
Comment (Jakuje) [inline src/objects.c:2438]: is this todo still valid? If so, I think it needs some more description as it is not clear to me what needs to be done and why.

File: src/objects.c [hunk 4]
```diff
@@ -1799,3 +1813,748 @@ int p11prov_obj_export_public_ec_key(P11PROV_OBJ *obj, OSSL_CALLBACK *cb_fn,
...
+    }
+
+    if (allocattrs < findctx.numattrs) {
+        allocattrs = findctx.numattrs;
+    }
+
+    /* A common case with openssl is the request to import a key we already
+     * actually have on the token. This happens because OpenSSL is greedy
+     * and tries to export keys to its default provider before it even knows
+     * wat kind of operation it needs to do. Sometimes the operation ends up
```
Comment (Jakuje) [inline src/objects.c:2508]:
```suggestion
     * what kind of operation it needs to do. Sometimes the operation ends up
```

File: src/objects.c [hunk 5]
```diff
@@ -1799,3 +1813,748 @@ int p11prov_obj_export_public_ec_key(P11PROV_OBJ *obj, OSSL_CALLBACK *cb_fn,
...
+        rv = return_dup_key(key, findctx.found);
+        goto done;
+    }
+
+    /*
+     * FIXME:
+     * For things like ECDH we can get away with a mock objects that just holds
+     * data for now, but is not backed by an actual handle and key in the token.
+     * Once this is not sufficient, we'll probably need to change functions to
+     * pass in a valid session when requesting a handle from an keyect, so that
```
Comment (Jakuje) [inline src/objects.c:2528]: `keyect`?

File: tests/tbasic [hunk 6]
```diff
@@ -150,7 +149,25 @@ if [ $FAIL -ne 0 ]; then
...
+title PARA "Test EVP_PKEY_eq on public EC key via import"
+$CHECKER ./tcmpkeys "$ECPUBURI" "${TMPPDIR}"/ec.pub.uripin.pem
+
+title PARA "Test EVP_PKEY_eq with key exporting disabled"
+ORIG_OPENSSL_CONF=${OPENSSL_CONF}
+OPENSSL_CONF=${OPENSSL_CONF}.noexport
+title PARA "Test RSA key"
+$CHECKER ./tcmpkeys "$PUBURI" "$PUBURI"
+title PARA "Test EC key"
+$CHECKER ./tcmpkeys "$ECPUBURI" "$ECPUBURI"
```
Comment (Jakuje) [inline tests/tbasic:169]: We should test also comparison of the EdDSA keys if possible

File: tests/tgenkey.c [hunk 7]
```diff
@@ -67,6 +68,49 @@ static void check_keys(OSSL_STORE_CTX *store)
         exit(EXIT_FAILURE);
     }
 
+    /* check we can get pub params from key */
```
Comment (Jakuje) [inline tests/tgenkey.c:71]: Same here, no coverage for the eddsa

File: src/objects.c [hunk 8]
```diff
@@ -1799,3 +1813,748 @@ int p11prov_obj_export_public_ec_key(P11PROV_OBJ *obj, OSSL_CALLBACK *cb_fn,
...
+        return rv;
+        ;
+    }
+    dst->numattrs = 0;
+    for (int i = 0; i < src->numattrs; i++) {
+        rv = p11prov_copy_attr(&dst->attrs[i], &src->attrs[i]);
+        if (rv != CKR_OK) {
+            rv = CKR_HOST_MEMORY;
+            P11PROV_raise(dst->ctx, rv, "Failed attr copy");
+            return rv;
```
[the remark this answers is not in the document]
Comment (simo5) [reply src/objects.c:2395]: no actually it is not needed, this is stored on the key object which will be freed if the whole thing fails, that is why I am careful to update dst->numattrs after each step, so the p11prov_obj_free() will deal with this correctly.

File: src/objects.c (hunk 5 above)
Comment (simo5) [reply src/objects.c:2528]: lol I did a s/obj/key/g at some point ... good catch

File: tests/tbasic (hunk 6 above)
Comment (simo5) [reply tests/tbasic:169]: maybe open an issue for that?

File: tests/tgenkey.c (hunk 7 above)
Comment (simo5) [reply tests/tgenkey.c:71]: yeah I think we should have a separate PR for that, this one grew enough

**Review by Jakuje — APPROVED** [review]
no other technical issue noticed so giving my ack

**Comment by simo5** [comment]
Thanks a lot @Jakuje
Given the only issues is this weird address sanitezer misbehavior, I am going to merge this now, and create new PRs for followups

File: tests/tgenkey.c (hunk 7 above)
Comment (simo5) [reply tests/tgenkey.c:71]: Great, thank you

**Pull request merged**